### PR TITLE
Add custom derives for specific generated types

### PR DESCRIPTION
**Pull request by ascjones**
Currently the user can only specify custom derives to be applied to _all_ generated types. Sometimes it is required to use a specific generated type with certain requirements e.g. `Ord, PartialOrd` or `serde::Serialize, serde::Deserialize`, however it is not always possible/desirable to implement those on all generated types. For example when used in conjuction with `substitute_type` where the substituted type does not implement the required derives.

This PR adds the ability to specify derives for specific generated types, by adding the `derive_for_type` configuration to the `subxt` attribute:

```
#[subxt::subxt(
    runtime_metadata_path = "examples/polkadot_metadata.scale",
    derive_for_all_types = "Clone, PartialEq, Hash",
    derive_for_type(ty = "sp_runtime::DispatchError", derive = "::serde::Deserialize") // todo check this syntax
)]
```
Note also that `generated_type_derives` has been renamed to `derive_for_all_types`.

The final set of derives applied to a specified type will be the union of `default_derives` + `user_derive_for_all_types` + `user_derive_for_type`.

## todo

- [x] example
- [x] unit tests
- [x] test with `cargo-contract`
- [x] docs

### CODE REVIEW

File: codegen/src/types/derives.rs [hunk 1]
```diff
@@ -17,48 +17,107 @@
...
+        derives: impl Iterator<Item = syn::Path>,
+    ) {
+        let type_derives = self
+            .specific_type_derives
+            .entry(ty)
+            .or_insert_with(Derives::default);
+        type_derives.derives.extend(derives)
+    }
+
+    /// Returns a the derives to be applied to all generated types.
```
Comment (lexnv) [inline codegen/src/types/derives.rs:53]:
```suggestion
    /// Returns the derives to be applied to all generated types.
```

File: codegen/src/types/mod.rs [hunk 2]
```diff
@@ -215,22 +219,31 @@ impl<'a> TypeGenerator<'a> {
         }
     }
 
-    /// Returns the derives with which all generated type will be decorated.
-    pub fn derives(&self) -> &GeneratedTypeDerives {
-        &self.derives
+    /// Returns a the derives to be applied to all generated types.
```
Comment (lexnv) [inline codegen/src/types/mod.rs:222]:
```suggestion
    /// Returns the derives to be applied to all generated types.
```

File: codegen/src/types/mod.rs [hunk 3]
```diff
@@ -215,22 +219,31 @@ impl<'a> TypeGenerator<'a> {
...
+    pub fn default_derives(&self) -> &Derives {
+        self.derives.default_derives()
+    }
+
+    /// Returns a the derives to be applied to a generated type.
```
Comment (lexnv) [inline codegen/src/types/mod.rs:227]:
```suggestion
    /// Returns the derives to be applied to a generated type.
```

**Review by lexnv — APPROVED** [review]

**Review by jsdw — APPROVED** [review]
This is great!

**Pull request merged**